### DISCUSSION TITLE
Multiple DataWriters with different IP Assignments in a Single DomainParticipant

### DISCUSSION THREAD

**Discussion by hejazifar** [eProsima/Fast-DDS#3844] [Support]
I have a notebook with two network interfaces, each having a different IP address. My goal is to configure two DataWriters within a single DomainParticipant, each sending its data through a specific IP address that I select. For instance, DataWriter1 should transmit data via IP1=192.168.0.1, and DataWriter2 should do so via IP2=192.168.0.2.

As far as I understood, for dataReader, it can be done by using the unique_network_flow feature. However, I could not find a solution for DataWriter. Is it possible to overwrite the locator list of dataWriter?

Currently, my workaround involves creating two separate DomainParticipants for each DataWriter and using a whitelist QoS for each DomainParticipant. However, this approach is neither efficient nor scalable, as eventually I will have multiple DataWriters and multiple interfaces.

I would greatly appreciate any assistance with this matter. My OS is Ubuntu 20.04

![Desire_vs_current](https://github.com/eProsima/Fast-DDS/assets/36082611/13cd4733-1020-4a0f-8adc-421fd9ab90cc)

**Comment by JLBuenoLopez** [comment]
Hi @hejazifar 

You can always use the [RTPSEndpointQos](https://fast-dds.docs.eprosima.com/en/latest/fastdds/dds_layer/core/policy/eprosimaExtensions.html#rtpsendpointqos) to set specific locators for each endpoint within the DomainParticipant. The [unique network flows feature](https://fast-dds.docs.eprosima.com/en/latest/fastdds/use_cases/unique_network_flows/unique_network_flows.html#requesting-unique-flows) helps with this configuration in the DataReader side, but the documentation already explains that this can also be achieved configuring properly the RTPSEndpointQos.

**Reply by hejazifar** [reply]
Hi @JLBuenoLopez-eProsima 

I appreciate your response. I have experimented with the RTPSEndpointQoS for both DataWriter and DataReader, but it hasn't behaved as expected. I have tested it using both an XML file and CPP code, and the results were similar. Below are the XML lines I have added within the <data_reader> or <data_writer> segment:

```xml
<unicastLocatorList>
        <locator>
            <udpv4>
                <address>10.45.0.1</address>
            </udpv4>
        </locator>
</unicastLocatorList>
```

To observe the behavior, I utilized the get_listening_locators() function for DataReader and get_sending_locators() for DataWriter after their creation. Subsequently, I printed the locators on the terminal and noticed the following:

- RTPSEndpointQoS solely restricts the listening locator on the DataReader and has no impact on the DataWriter sending locators.
- After the discovery phase and a match between the DataWriter and DataReader, if I disable the IP interface (in this case, 10.45.0.1) on the DataReader side, the DataReader still receives data from other interfaces.

This raises a few questions:

- Can you confirm if RTPSEndpointQoS has any effect on the DataWriter? I'm uncertain whether the RTPSEndpointQoS is only related to DataReader, as the documentation stated that the unique network flow feature is exclusively available for DataReader.
- Why does the DataReader continue to receive data when I disable the interface associated with the IP address specified in RTPSEndpointQoS?

I would greatly appreciate any insights you can provide.
Thank you in advance
